File: src/srv-alarm/src/ringer.rs
use futures_util::FutureExt;
use tokio::sync::mpsc;
use tokio::time::{self, Duration, Instant};
use tokio::{join, select};

use crate::http;
use crate::manager::Action as MgrAction;
use crate::models::{Context, Status};

pub type Ringer = mpsc::UnboundedSender<Action>;

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum Action {
    Start,
    Stop,
}

const TICK_DURATION: Duration = Duration::from_secs(60);

pub fn start(ctx: Context) -> Ringer {
    let (tx, mut rx) = mpsc::unbounded_channel::<Action>();
    tokio::spawn(async move {
        let mut delay_next_step = Box::pin(time::sleep_until(Instant::now()).fuse());
        let mut minute = 0;

        loop {
            select! {
                action = rx.recv() => {
                    if action.is_none() {
                        log::error!("All ringer radio senders closed. Exiting ringer ...");
                        break;
                    }

                    let duration = handle_action(&ctx, action.unwrap()).await;
                    if let Some(duration) = duration {
                        minute = 0;
                        log::info!("Reset delay to {:?}s", duration.as_secs());
                        delay_next_step = Box::pin(time::sleep_until(Instant::now() + duration).fuse());
                    }
                }

                _ = &mut delay_next_step => {
                    // Skip if alarm is idle
                    minute += 1;
                    if ctx.get_status() != Status::Idle {
                        log::info!("Handle next step");
                        handle_next_step(&ctx, minute).await;
                        log::info!("Reset delay to {}s", TICK_DURATION.as_secs());
                        delay_next_step = Box::pin(time::sleep_until(Instant::now() + TICK_DURATION).fuse());
                    } else {
                        log::info!("Next step skipped since status is idle")
                    }
                }
            }
        }
    });
    return tx;
}

async fn handle_action(ctx: &Context, action: Action) -> Option<Duration> {
    log::info!("Handle action: {:?}", action);
    match action {
        Action::Start => handle_start(ctx).await,
        Action::Stop => handle_stop(ctx).await,
    }
}

async fn handle_start(ctx: &Context) -> Option<Duration> {
    // Start light
    log::info!("Starting sunrise");
    http::start_sunrise(ctx).await.ok();
    Some(TICK_DURATION)
}

async fn handle_stop(ctx: &Context) -> Option<Duration> {
    // Stop light and sound
    log::info!("Stopping sunrise");
    log::info!("Stopping alarm music");
    log::info!("Stopping buzzer");
    let (leds, music, buzzer) = join!(
        http::stop_sunrise(ctx),
        http::stop_music(ctx),
        http::stop_buzzer(ctx)
    );
    leds.ok();
    music.ok();
    buzzer.ok();

    // Don't set next step
    None
}

async fn handle_next_step(ctx: &Context, minute: u8) {
    // General
    let minute = minute as i64;
    log::info!("Handle next alarm step at minute: {}", minute);

    // Check for abort
    let light_duration = ctx.config.alarm.light_duration.num_minutes();
    let abort_delay = light_duration + 10;
    if minute > abort_delay {
        log::warn!("Alarm reached abort limit. Requesting manager to abort alarm.");
        ctx.radio
            .send(MgrAction::AbortAlarm)
            .map_err(|e| log::error!("Failed to send action AbortAlarm to manager: {}", e))
            .ok();
        return;
    }

    // Check for buzzer
    if minute >= light_duration {
        log::info!("Starting buzzer");
        http::start_buzzer(ctx).await.ok();
    }

    // Check for music
    let sound_delay = ctx.config.alarm.light_duration - ctx.config.alarm.sound_duration;
    if minute == sound_delay.num_minutes() {
        log::info!("Starting alarm music");
        http::start_music(ctx).await.ok();
    } else if minute > sound_delay.num_minutes() {
        log::info!("Increasing alarm volume");
        http::increase_music_volume(ctx).await.ok();
    }
}

File: src/services/alarm/usecases/ringer.go
package usecases

import (
	"sync"
	"time"

	"github.com/JenswBE/sunrise-alarm/src/services/audio"
	"github.com/JenswBE/sunrise-alarm/src/services/physical"
	"github.com/JenswBE/sunrise-alarm/src/utils/trigger"
	"github.com/rs/zerolog/log"
	"github.com/samber/lo"
)

const (
	lightDelay    = 0 * time.Minute
	lightDuration = 10 * time.Minute
	lightDone     = lightDelay + lightDuration

	soundDelay    = 3 * time.Minute
	soundDuration = 7 * time.Minute
	soundDone     = soundDelay + soundDuration

	buzzerDelay    = 10 * time.Minute
	buzzerDuration = 5 * time.Minute
	buzzerDone     = buzzerDelay + buzzerDuration

	tickDuration = time.Minute
)

type Ringer struct {
	ringingMutex        sync.Mutex
	ringing             bool
	ringingChan         chan bool
	ringingSinceMinutes uint
	lightStatus         stepStatus
	soundStatus         stepStatus
	buzzerStatus        stepStatus
	abortAlarm          chan<- struct{}
	audioService        audio.Service
	physicalService     physical.Service
}

type stepStatus string

const (
	stepStatusAwaitingStart stepStatus = "AWAITING_START"
	stepStatusOngoing       stepStatus = "ONGOING"
	stepStatusDone          stepStatus = "DONE"
)

func NewRinger(physicalService physical.Service, audioService audio.Service, abortAlarm chan<- struct{}) *Ringer {
	// Init
	ringer := &Ringer{
		ringingChan:     make(chan bool, 1),
		abortAlarm:      abortAlarm,
		audioService:    audioService,
		physicalService: physicalService,
	}

	// Start event loop
	go ringer.eventLoop(abortAlarm)
	return ringer
}

func (r *Ringer) eventLoop(abortAlarm chan<- struct{}) {
	trigger := trigger.NewDelayedTrigger()
	for {
		select {
		case shouldRing := <-r.ringingChan:
			if shouldRing {
				r.ringingSinceMinutes = 0
				r.lightStatus = stepStatusAwaitingStart
				r.soundStatus = stepStatusAwaitingStart
				r.buzzerStatus = stepStatusAwaitingStart
				r.handleNextStep()
				trigger.Schedule(tickDuration)
			} else {
				if r.lightStatus == stepStatusOngoing {
					log.Debug().Msg("Ringer.handleNextStep: Stopping alarm light because alarm is stopped/aborted")
					r.physicalService.StopSunriseSimulation()
				}
				if r.soundStatus == stepStatusOngoing {
					log.Debug().Msg("Ringer.handleNextStep: Stopping alarm sound because alarm is stopped/aborted")
					if err := r.audioService.StopMusic(); err != nil {
						log.Error().Err(err).Msg("Ringer: Failed to stop music")
					}
				}
				if r.buzzerStatus == stepStatusOngoing {
					log.Debug().Msg("Ringer.handleNextStep: Stopping alarm buzzer because alarm is stopped/aborted")
					r.physicalService.StopBuzzer()
				}
			}
		case <-trigger.C:
			if !r.ringing {
				log.Debug().Msg("Ringer.loop: Received delayed trigger, but ringer is not ringing. Ignoring delayed trigger.")
				continue
			}
			r.ringingSinceMinutes++
			r.handleNextStep()
			trigger.Schedule(tickDuration)
		}
	}
}

func (r *Ringer) handleNextStep() {
	// Init
	currentDelay := time.Duration(r.ringingSinceMinutes) * time.Minute
	logger := log.With().Stringer("current_delay", currentDelay).Logger()
	logger.Debug().Msgf("Ringer.handleNextStep: Handle next alarm step at minute %d", r.ringingSinceMinutes)

	// Check for abort
	abortDelay := lo.Max([]time.Duration{lightDone, soundDone, buzzerDone})
	if currentDelay >= abortDelay {
		logger.Warn().Stringer("abort_delay", abortDelay).Msg("Ringer.handleNextStep: Ringer reached abort limit. Requesting manager to abort alarm.")
		r.abortAlarm <- struct{}{}
		return
	}

	// Check for light
	switch {
	case currentDelay >= lightDelay && r.lightStatus == stepStatusAwaitingStart:
		logger.Debug().Msg("Ringer.handleNextStep: Starting alarm light")
		r.physicalService.StartSunriseSimulation()
		r.lightStatus = stepStatusOngoing
	case currentDelay >= lightDone && r.lightStatus == stepStatusOngoing:
		logger.Debug().Msg("Ringer.handleNextStep: Stopping alarm light because done is reached")
		r.physicalService.StopSunriseSimulation()
		r.lightStatus = stepStatusDone
	}

	// Check for sound
	switch {
	case currentDelay >= soundDelay && r.soundStatus == stepStatusAwaitingStart:
		logger.Debug().Msg("Ringer.handleNextStep: Starting alarm sound")
		if err := r.audioService.PlayMusic(); err != nil {
			logger.Error().Msg("Ringer.handleNextStep: Failed to play music")
		}
		r.soundStatus = stepStatusOngoing
	case currentDelay < soundDone && r.soundStatus == stepStatusOngoing:
		logger.Debug().Msg("Ringer.handleNextStep: Increasing alarm sound volume")
		if err := r.audioService.IncreaseVolume(); err != nil {
			logger.Error().Msg("Ringer.handleNextStep: Failed to increase alarm sound volume")
		}
	case currentDelay >= soundDone && r.soundStatus == stepStatusOngoing:
		logger.Debug().Msg("Ringer.handleNextStep: Stopping alarm sound because done is reached")
		if err := r.audioService.StopMusic(); err != nil {
			log.Error().Err(err).Msg("Ringer: Failed to stop music")
		}
		r.soundStatus = stepStatusDone
	}

	// Check for buzzer
	switch {
	case currentDelay >= buzzerDelay && r.buzzerStatus == stepStatusAwaitingStart:
		logger.Debug().Msg("Ringer.handleNextStep: Starting alarm buzzer")
		r.physicalService.StartBuzzer()
		r.buzzerStatus = stepStatusOngoing
	case currentDelay >= buzzerDone && r.buzzerStatus == stepStatusOngoing:
		logger.Debug().Msg("Ringer.handleNextStep: Stopping alarm buzzer because done is reached")
		r.physicalService.StopBuzzer()
		r.buzzerStatus = stepStatusDone
	}
}

func (r *Ringer) Start() {
	log.Debug().Msg("Ringer.Start called")
	r.setRinging(true)
}

func (r *Ringer) Stop() {
	log.Debug().Msg("Ringer.Stop called")
	r.setRinging(false)
}

func (r *Ringer) setRinging(value bool) {
	r.ringingMutex.Lock()
	defer r.ringingMutex.Unlock()
	if r.ringing != value {
		r.ringing = value
		r.ringingChan <- value
	}
}
